src/main/java/reactor/core/publisher/FluxPeekFuseable.java

```diff
@@ -258,7 +263,7 @@ public T poll() {
 					throw Exceptions.propagate(Operators.onOperatorError(s, e, v));
 				}
 			}
-			if (v == null && sourceMode == SYNC) {
+			if (v == null && (d || sourceMode == SYNC)) {
```

Comment:
I'd reverse the condition to only touch volatile if necessary if we go that route